pkg/io/v2/opener.go

```diff
@@ -0,0 +1,135 @@
+type Opener interface {
+	Reader(ctx context.Context, path string, opts *blob.ReaderOptions) (io.ReadCloser, error)
+	Writer(ctx context.Context, path string, opts *blob.WriterOptions) (io.WriteCloser, error)
+}
+
+type opener struct {
+	credentials        []byte
+	cachedBuckets      map[string]*blob.Bucket
+	cachedBucketsMutex sync.Mutex
+}
+
+// NewOpener returns an opener that can read GCS, S3 and local paths.
+func NewOpener(credentialsFile string) (Opener, error) {
+	var credentials []byte
+	var err error
+	if credentialsFile != "" {
+		credentials, err = ioutil.ReadFile(credentialsFile)
+		if err != nil {
+			return nil, err
+		}
+	}
+	return &opener{
+		credentials:   credentials,
+		cachedBuckets: map[string]*blob.Bucket{},
+	}, nil
+}
+
+// getBucket opens a bucket
+// The storageProvider is discovered based on the given path.
+// The buckets are cached per bucket name. So we don't open a bucket multiple times in the same process
+func (o *opener) getBucket(ctx context.Context, path string) (*blob.Bucket, string, error) {
+	_, bucketName, relativePath, err := providers.ParseStoragePath(path)
+	if err != nil {
+		return nil, "", fmt.Errorf("could not get bucket: %w", err)
+	}
+
+	o.cachedBucketsMutex.Lock()
+	defer o.cachedBucketsMutex.Unlock()
+	if bucket, ok := o.cachedBuckets[bucketName]; ok {
+		return bucket, relativePath, nil
+	}
+
+	bucket, err := providers.GetBucket(ctx, o.credentials, path)
+	if err != nil {
+		return nil, "", err
+	}
+	o.cachedBuckets[bucketName] = bucket
+	return bucket, relativePath, nil
+}
+
+// Reader opens a reader on the given path
+// Internally a bucket is opened and cached. The storageProvider is discovered based on the path.
+func (o *opener) Reader(ctx context.Context, path string, opts *blob.ReaderOptions) (io.ReadCloser, error) {
+	bucket, relativePath, err := o.getBucket(ctx, path)
+	if err != nil {
+		return nil, err
+	}
+	reader, err := bucket.NewReader(ctx, relativePath, opts)
+	if err != nil {
+		return nil, err
+	}
+	return reader, nil
+}
+
+// Writer opens a writer on the given path
+// Internally a bucket is opened and cached. The storageProvider is discovered based on the path.
+func (o *opener) Writer(ctx context.Context, path string, opts *blob.WriterOptions) (io.WriteCloser, error) {
+	bucket, relativePath, err := o.getBucket(ctx, path)
+	if err != nil {
+		return nil, err
+	}
+	writer, err := bucket.NewWriter(ctx, relativePath, opts)
+	if err != nil {
+		return nil, err
+	}
+	return writer, nil
+}
+
+// ErrNotFoundTest can be used for unit tests to simulate NotFound errors.
+// This is required because gocloud doesn't expose its errors.
+var ErrNotFoundTest = fmt.Errorf("not found error which should only be used in tests")
+
+// IsNotExist will return true if the error shows that the object does not exist.
+func IsNotExist(err error) bool {
+	if err == ErrNotFoundTest {
+		return true
+	}
+	return gcerrors.Code(err) == gcerrors.NotFound
```

Comment:
This also needs to return true if err is or contains(errors.As) an os.IsNotExist

Reply:
I added the following. Guess errors.Is is also enough and you meant os.ErrNotExist:
```
	if errors.Is(err, os.ErrNotExist) {
		return true
	}
```